### PR TITLE
[rllib] Develop MARWIL

### PR DESCRIPTION
**Pull request by joneswong**
## What do these changes do?
To implement MARWIL (see their [neurips18 paper](http://papers.nips.cc/paper/7866-exponentially-weighted-imitation-learning-for-batched-historical-data)). MARWIL learns a policy from hitorical data generated by unknown policies.

- [x] create computation graph
- [x] create corresponding agent class (mainly process the common outputs of other rllib agents, from immediate reward to discounted accumulated rewards)
- [x] solve some benchmark games with DQN, PPO, ES, etc., and Learn MARWIL from their experience.

## Related issue number
#3607

### CODE REVIEW

File: python/ray/rllib/optimizers/offline_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                batch_indexes = -np.ones_like(rewards)
+                samples[policy_id] = SampleBatch({
+                    "obs": obses_t,
+                    "actions": actions,
+                    "rewards": rewards,
+                    "new_obs": obses_tp1,
+                    "dones": dones,
+                    "weights": weights,
+                    "batch_indexes": batch_indexes
+                })
```
Comment (ericl) [inline python/ray/rllib/optimizers/offline_optimizer.py]: Is it possible to use the input API and SyncReplayOptimizer or even SyncSamplesOptimizer instead of this? The functionality looks quite similar.

Comment (ericl) [reply python/ray/rllib/optimizers/offline_optimizer.py]: Hmm, I guess the problem is that in the input API right now, postprocess_trajectory() is not called on input batches read from files. So there is no way to compute advantages for the batch if it's not already present. If we allowed this, would that resolve the issue?

Comment (joneswong) [reply python/ray/rllib/optimizers/offline_optimizer.py]: Happy new year! I just came back. I think `postprocess_trajectory()` works if its `sample_batch` argument is fed with a complete trajectory. In my current implementation, I cache the prefix of current trajectory and compute the advantages once a `done` is encountered. To use `SyncReplayOptimizer` or `SyncSamplesOptimizer` instead, where do you suggest to call `postprocess_trajectory()`? Thanks.

**Comment by ericl** [comment]
Happy new year too!

I think we can assume the complete episode is present in the batch
("complete_episodes" batch mode option) and use vf estimation if not.

I guess you can use the replay optimizer since it is step oriented, but
sync samples or batch replay should work:

https://github.com/ray-project/ray/blob/master/python/ray/rllib/optimizers/sync_batch_replay_optimizer.py

File: python/ray/rllib/optimizers/offline_optimizer.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/optimizers/offline_optimizer.py]: You can set the "postprocess_inputs" option here: https://ray.readthedocs.io/en/latest/rllib-offline.html#input-api

**Comment by joneswong** [comment]
train on the experience generated during the training course of a PPO:
cartpole-marwil

**Comment by ericl** [comment]
Interesting, presumably you don't reach 200 since you're imitating the bad experiences as well?

**Comment by joneswong** [comment]
> Interesting, presumably you don't reach 200 since you're imitating the bad experiences as well, but it should do better than vanilla IL?

learning starts until there are 300k samples in the buffer. The PPO reaches 200 consuming around 400k and I actually generated 1000k. It is reasonable to compare with imitation learning. It seems that I can implement IL in the MARWILPolicyGraph as well.

**Comment by joneswong** [comment]
cartpole

File: python/ray/rllib/agents/marwil/marwil_policy_graph.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+            self.loss = -1.0 * tf.reduce_mean(tf.stop_gradient(adv) * logprobs)
+
+
+class MARWILPolicyGraph(TFPolicyGraph):
+    def __init__(self, observation_space, action_space, config):
+        config = dict(ray.rllib.agents.dqn.dqn.DEFAULT_CONFIG, **config)
+        # TO DO: support hybrid action
+        if not isinstance(action_space, Discrete):
+            raise UnsupportedSpaceException(
```
Comment (ericl) [inline python/ray/rllib/agents/marwil/marwil_policy_graph.py]: This should work with all spaces if you use ModelCatalog.get_action_dist() right?

File: python/ray/rllib/agents/marwil/marwil.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    # How many steps of the model to sample before learning starts.
+    "learning_starts": 50000,
+
+    # === Parallelism ===
+    # Number of workers for collecting samples with. This only makes sense
+    # to increase if your environment is particularly slow to sample, or if
+    # you"re using the Async or Ape-X optimizers.
+    "num_workers": 0,
+    # Optimizer class to use.
+    "optimizer_class": "SyncReplayOptimizer",
```
Comment (ericl) [inline python/ray/rllib/agents/marwil/marwil.py]: Wouldn't it be better to use SyncSamplesOptimizer with no replay? The input reader will already loop over inputs.

File: python/ray/rllib/agents/marwil/marwil_policy_graph.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+from ray.rllib.evaluation.policy_graph import PolicyGraph
+from ray.rllib.evaluation.tf_policy_graph import TFPolicyGraph
+from ray.rllib.agents.dqn.dqn_policy_graph import _scope_vars
+
+P_SCOPE = "p_func"
+V_SCOPE = "v_func"
+
+
+class PNetwork(object):
+    def __init__(self, model, num_actions, hiddens=[256]):
```
Comment (ericl) [inline python/ray/rllib/agents/marwil/marwil_policy_graph.py]: Rename to PolicyNetwork for readability?

File: python/ray/rllib/agents/marwil/marwil.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+# __sphinx_doc_begin__
+DEFAULT_CONFIG = with_common_config({
+    # If true, use the Generalized Advantage Estimator (GAE)
+    # with a value function, see https://arxiv.org/pdf/1506.02438.pdf.
+    "use_gae": False,
+    # GAE(lambda) parameter
+    "lambda": 1.0,
+
+    # === Model ===
+    "actor_hiddens": [64],
```
Comment (ericl) [inline python/ray/rllib/agents/marwil/marwil.py]: I think we should avoid having these types of options, are the model options sufficient?

**Comment by joneswong** [comment]
I revised accordingly and fixed incorrectly usage of input data (use reward rather than advantage.....very sorry for such a  stupid mistake). I generated historic data via a PPO agent (first, train a PPO agent until cartpole episode reward mean = 100, then restore from the checkpoint and evaluate for around 1M steps with learning rate=0). Then I trained IL and MARWIL on this data (each for 3 trials). The results are as follow:
cartpole

File: python/ray/rllib/agents/marwil/marwil_policy_graph.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        with tf.control_dependencies([update_adv_norm]):
+            exp_advs = tf.exp(
+                beta * tf.divide(adv, 1e-8 + tf.sqrt(ma_adv_norm)))
+
+        # log\pi_\theta(a|s)
+        dist_cls, _ = ModelCatalog.get_action_dist(action_space, {})
+        action_dist = dist_cls(logits)
+        logprobs = action_dist.logp(actions)
+
+        self.loss = -1.0 * tf.reduce_mean(tf.stop_gradient(adv) * logprobs)
```
Comment (ericl) [inline python/ray/rllib/agents/marwil/marwil_policy_graph.py]: Shouldn't this be against `exp_advs` not `adv`?

**Review by ericl — APPROVED** [review]
@joneswong I made some minor revisions, fixed the `exp_adv` issue. I also switched to the batch replay optimizer so that the samples are shuffled instead of read in order.

If this looks fine to you, we can merge.

File: python/ray/rllib/agents/marwil/marwil_policy_graph.py (hunk 6 above)
Comment (joneswong) [reply python/ray/rllib/agents/marwil/marwil_policy_graph.py]: oh...my fault. Thanks!

Comment (joneswong) [reply python/ray/rllib/agents/marwil/marwil_policy_graph.py]: It very critical when the historical data include trajectories generated by different policies in which case the advantage weights for a certain (s, a) pair may vary a lot and exponential op makes them non-negative. I am very sorry. Thank you!

**Comment by joneswong** [comment]
sure, it also looks good to me.

**Comment by zzcNEU** [comment]
hello, I try to use the marwil to train the model cartpole. But, it remainds me that rewads is NAN
